core/src/main/java/org/apache/iceberg/DeltaLakeDataTypeVisitor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import io.delta.standalone.types.ArrayType;
+import io.delta.standalone.types.DataType;
+import io.delta.standalone.types.MapType;
+import io.delta.standalone.types.StructField;
+import io.delta.standalone.types.StructType;
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class DeltaLakeDataTypeVisitor<T> {
+  public static <T> T visit(DataType type, DeltaLakeDataTypeVisitor<T> visitor) {
```

Comment:
This class, as well as `DeltaLakeTypeToType`, is derived from [SparkTypeToType](https://github.com/apache/iceberg/blob/be0c8ddfd2a6d3c0f2e38b0b980d8fe73980ca3f/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkTypeToType.java) and [SparkTypeVisitor](https://github.com/apache/iceberg/blob/be0c8ddfd2a6d3c0f2e38b0b980d8fe73980ca3f/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkTypeVisitor.java). It aims to perform a direct transformation from Delta Lake's schema to Iceberg's schema.